src/api.js

```diff
@@ -1,46 +1,70 @@
-const rest = require(`./rest`);
-const ws = require(`./ws`);
-let type;
-
-/**
- * Initialize API based on protocol and URL
- * 
- * @param {String} serverURL 
- * @returns 
- */
-function init(serverURL){
-  type = serverURL.slice(0, 2);
-  if (type === `ht`){
-    return rest.init(serverURL);
-  } else {
-    return ws.init(serverURL);
-  }
-}
+'use strict';
 
-/**
- * Returns api type
- * @returns {*}
- */
-function getType(){
-  return type;
-}
+// Requirements
+
+const Rest = require(`./rest`);
+const WS = require(`./ws`);
+
+
+// Source
+
+class API {
+    /**
+     * @param {Object} options { authServiceURL, mainServiceURL, pluginServiceURL }
+     */
+    constructor(options) {
+        const { authServiceURL, mainServiceURL, pluginServiceURL } = options;
+
+        const SelectedAPI = this.getType(mainServiceURL);
 
-/**
- * Returns current API
- * 
- * @returns 
- */
-function getAPI(){
-  if (type === `ht`){
-    return rest;
-  } else {
-    return ws;
-  }
+        if (SelectedAPI) {
+            this.schema = new SelectedAPI({
+                authServiceURL,
+                mainServiceURL,
+                pluginServiceURL
+            });
+        } else {
+            throw new Error('unexpected mainServiceURL, please use allowed protocol');
+        }
+
+        return this.schema;
+    }
+
+    /**
+     * Returns current API
+     * 
+     * @returns 
+     */
+    getType(url) {
+        let parsedURL;
+        if (url) parsedURL = url.match(/^([a-z]{1,5}):\/\//);
```

Comment:
Using of regexp to determinate protocol has lots of overheads. Would it be better to use String.startsWith method? Example:

url.toLowerCase().startsWith(`http`)

Reply:
@itrambovetskyi when there will be also 'mqtt' protocol we need to make startsWith 3 times. Is it a good idea?

Reply:
In the current code, in the switch below, you compare string. It is almost the same as startsWidth. So, for now you have 1 regexp and 3 startsWith.

Reply:
OK, will remake.